DEVX-2723: call confluent kafka cluster use before stack tear-down

javabrett opened the pull request:
### Description 

https://confluentinc.atlassian.net/browse/DEVX-2723

_What behavior does this PR change, and why?_

Specifies `--cluster` ID when running `confluent kafka acl delete` during tear-down, to prevent attempt to delete ACL from the context cluster, should a cluster different from the demo cluster be selected when `stop-cloud.sh` is run.

### Author Validation

_Describe the validation already done, or needs to be done, by the PR submitter._

<!-- Uncomment any of the following that are required -->
<!-- - [ ] Documentation -->
<!-- - [ ] ccloud/beginner-cloud -->
 - [X] ccloud/ccloud-stack
<!-- - [ ] ccloud-observability -->
<!-- - [ ] clickstream -->
<!-- - [ ] clients/avro -->
<!-- - [ ] clients/cloud -->
<!-- - [ ] cloud-etl -->
<!-- - [ ] connect-streams-pipeline -->
<!-- - [ ] cp-quickstart -->
<!-- - [ ] kubernetes/gke-base -->
<!-- - [ ] kubernetes/replicator-gke-cc -->
<!-- - [ ] microservices-orders -->
<!-- - [ ] multi-datacenter -->
<!-- - [ ] multiregion -->
<!-- - [ ] replicator-schema-translation -->
<!-- - [ ] replicator-security -->
<!-- - [ ] security/rbac -->


### Reviewer Tasks

_Describe the tasks/validation that the PR submitter is requesting to be done by the reviewer._

<!-- Uncomment any of the following that are required -->
<!-- - [ ] Documentation -->
<!-- - [ ] ccloud/beginner-cloud -->
<!-- - [ ] ccloud/ccloud-stack -->
<!-- - [ ] ccloud-observability -->
<!-- - [ ] clickstream -->
<!-- - [ ] clients/avro -->
<!-- - [ ] clients/cloud -->
<!-- - [ ] cloud-etl -->
<!-- - [ ] connect-streams-pipeline -->
<!-- - [ ] cp-quickstart -->
<!-- - [ ] kubernetes/gke-base -->
<!-- - [ ] kubernetes/replicator-gke-cc -->
<!-- - [ ] microservices-orders -->
<!-- - [ ] multi-datacenter -->
<!-- - [ ] multiregion -->
<!-- - [ ] replicator-schema-translation -->
<!-- - [ ] replicator-security -->
<!-- - [ ] security/rbac -->

javabrett committed:
DEVX-2723: specify --cluster (ID) with confluent kafka acl delete.

ybyzek commented:
Per Slack, instead of doing this PR, the fix could be to `use` the Kafka cluster ID identified

javabrett committed:
DEVX-2724: specify --cluster (ID) with confluent connect delete.

javabrett committed:
DEVX-2723/DEVX-2724: always 'confluent kafka cluster use <clusterId>' before tear-down.

javabrett commented:
Added two commits to:
- Address DEVX-2724 also, by adding `--cluster` to `confluent connect delete` calls during tear-down
- Added call to `confluent kafka cluster use` prior to teardown: good hygiene to ensure the correct cluster is selected prior to tear-down.

ybyzek commented on utils/ccloud_library.sh:
```diff
@@ -1055,8 +1056,11 @@ function ccloud::destroy_ccloud_stack() {
   echo "Destroying Confluent Cloud stack associated to service account id $SERVICE_ACCOUNT_ID"
 
+  local cluster_id=$(confluent kafka cluster list -o json | jq -r 'map(select(.name == "'"$CLUSTER_NAME"'")) | .[].id')
+  confluent kafka cluster use $cluster_id 2>/dev/null
```
Given this statement, is it possible that the other `--cluster $CLUSTER_ID` are extraneous and can be removed?

javabrett replied:
I decided both `--cluster` instances were safest and correct to retain, albeit unnecessary following successful `confluent kafka cluster use`.  Ideally we don't rely on this context for any commands - ideally all the commands are independent of context by fully specifying context via args.

But no, they aren't necessary, if removed I expect the bug would remain fixed.

ybyzek replied:
For simplicity (and therefore maintenance0, my vote is to keep this PR to just these 2 lines here

```
  local cluster_id=$(confluent kafka cluster list -o json | jq -r 'map(select(.name == "'"$CLUSTER_NAME"'")) | .[].id')
   confluent kafka cluster use $cluster_id 2>/dev/null
```

And remove the redundancy of the other lines in this PR (`--cluster $CLUSTER_ID`)

javabrett replied:
@ybyzek how would you feel about instead retaining the `--cluster` args over the `cluster use`?

That way there is no ambiguity in the commands, they are explicit and precise about what they select for deletion - if for example a user were able to fluke an accidental concurrent `confluent` command use in another shell, these deletes would remain correct and work even if the `cluster use` has been undone, or otherwise failed.

ybyzek replied:
I understand the goal but there are so many other places in this ccloud library (aside from this function) where concurrent `confluent` commands can result in problems, that IMO it's safer to not set that expectation with the user.  It could be a separate effort to harden those functions to meet that expectation.  For now, given what little cycles we have for `ccloud-stack` in general, IMO it's better to optimize for simpler code.

javabrett committed:
Don't pass --cluster to delete commands.

This commit partially reverts commits:

- 793c168414ad8ba022d75a1169d91c2dc766548a
- 4ef85069ae64cef480b3fdbb3a851d5f8dc89f60

davetroiano reviewed (approved):
LGTM!
